src/par/FixColumnPartitions.c

```diff
@@ -19,138 +19,162 @@
+  int found;
+  List_ptr rfaces, rfaces2;
+  int nrf, i, j, k, nfv, ret;
+  MFace_ptr rf; double rcen[3], upcen[3], dncen[3];
+  double fxyz[MAXPV2][3];
+  MRegion_ptr mr_it;
+
+  *up=NULL; *dn=NULL;
+  
+  rfaces = MR_Faces(mr);
+  nrf = List_Num_Entries(rfaces);
+
+  /* find two faces that are not sides */
+  found = 0; i = 0;
+  while (found < 2) {
+    rf = List_Entry(rfaces,i);
+    if (!FixColumnPartitions_IsSide(mesh, mr, rf)) {
+      if (found < 1) *up = rf;
+      else *dn = rf;
+      found++;
     }
+    i++;
+    if (i >= nrf) 
+      MSTK_Report("FixColumnPartitions","Mesh is not columnar, can't find both up and down faces.",MSTK_FATAL);
+  }
+
+  /* figure which is up and which is down */
+  MF_Coords(*up,&nfv,fxyz);
+  upcen[0] = upcen[1] = upcen[2] = 0.;
+  for (j = 0; j < nfv; j++) {
     for (k = 0; k < 3; k++)
-      rcen[k] /= nrv;        
-
-    topf = NULL;
-    maxdp = -1.0e+10;
-    for (i = 0; i < nrf; i++) {
-      rf = List_Entry(rfaces,i);
-      MF_Coords(rf,&nfv,fxyz);
-
-      fcen[0] = fcen[1] = fcen[2] = 0.0;
-      for (j = 0; j < nfv; j++) {
-        for (k = 0; k < 3; k++)
-          fcen[k] += fxyz[j][k];
-      }
-      for (k = 0; k < 3; k++)
-        fcen[k] /= nfv;
-
-      for (k = 0; k < 3; k++)
-        rfvec[k] = fcen[k]-rcen[k];
-      norm = sqrt(rfvec[0]*rfvec[0]+rfvec[1]*rfvec[1]+rfvec[2]*rfvec[2]);
-      for (k = 0; k < 3; k++)
-        rfvec[k] /= norm;
-
-      /* rfvec[2] is same as rfvec dotted with [0,0,1] */
-      if (rfvec[2] > maxdp) {
-        maxdp = rfvec[2];
-        topf = rf;
-      }
-    }
-    List_Delete(rfaces);
+      upcen[k] += fxyz[j][k];
+  }
 
-    if (!topf)
-      MSTK_Report("FixColumnPartitions","Could not find top face?",MSTK_FATAL);
+  MF_Coords(*dn,&nfv,fxyz);
+  dncen[0] = dncen[1] = dncen[2] = 0.;
+  for (j = 0; j < nfv; j++) {
+    for (k = 0; k < 3; k++)
+      dncen[k] += fxyz[j][k];
+  }
 
-    /* Check if the top face is also a top boundary face */
+  /* ASSERT x,y are the same */
+  if (fabs(upcen[0] - dncen[0]) > 1.e-12 ||
+      fabs(upcen[1] - dncen[1]) > 1.e-12) {
+    MSTK_Report("FixColumnPartitions","Mesh is not quite columnar, up and down faces do not align.",MSTK_FATAL);
+  }
 
-    fregs = MF_Regions(topf);
-    nfr = List_Num_Entries(fregs);
+  if (upcen[2] < dncen[2] - 1.e-12) {
+    rf = *up;
+    *up = *dn;
+    *dn = rf;
+    ret = 1;
+  } else if (upcen[2] > dncen[2] + 1.e-12) {
+    ret = 1;
+  } else {
+    /* find the first region upward that is proper */
+    MSTK_Report("FixColumnPartitions","Partitioning with degenerate cells is not yet supported.",MSTK_FATAL);
+    ret = 0;
+  }
 
-    if (nfr != 1) {
-      List_Delete(fregs);
-      continue; /* Not a mesh region at the top of the mesh */
-    }
-    
-    /* Hopefully, found the top face of a column in the mesh */
+  List_Delete(rfaces);
+  return ret;
+}
 
-    curreg = List_Entry(fregs,0);
-    curid = MR_ID(curreg);
-    homepid = part[curid-1];
+  
```

Comment:
There is some repetition of work in this routine (as there was in the previous version). If this is an issue, one can go through all faces and mark them as being "side" faces or not (using a temporary array) and then use that to zoom into the up and down faces in the FixColumnPartions_UpDown